### ISSUE TITLE
BERTopic Visualisations

### ISSUE THREAD

**Issue by clarkegavin** [clarkegavin/political_discourse#110]
Add additional visualisations to BERTopic
https://maartengr.github.io/BERTopic/getting_started/visualization/visualization.html#visualize-documents

**1. Visualise documents (visualize_documents) -> useful comparision between two datasets, but then only do this when we have selected the final model**
2. Visualise Topic Hierarchy (visualize_hierarchy) -> can help to select an appropriate nr_topics -> not sure how relevant thsi will be...
**3. Hierarchical Labels (hierarchical_topics -> visualize_hierarchy -> this is probably useful...**
4. Text-based topic tree (get_topic_tree)
**5. visualize hierarchical documents (hierarchical_topics -> visualize_hierarchical_documents) -> this could be useful as a comparison**
**6. Visualize terms (visualize_barchart) -> this is useful, again for comparison**
**7. Visualize topic similarity (visualize_heatmap()) - this is useful**
8. Visualize term score decline (visualize_term_rank) - pass log_scale=True)
**9. Visualize topics over time (topics_over_time) -> this is useful**
**10. visualize topics per class (useful for TDs, Constituencies, Parties) (visualize_topics_per_class)  - this is useful**
11. visualize probabilities or distribution (approximate_distribution

------
1. However, you might want a more fine-grained approach where we can visualize the documents inside the topics to see if they were assigned correctly or whether they make sense. To do so, we can use the topic_model.visualize_documents() function. This function recalculates the document embeddings and reduces them to 2-dimensional space for easier visualization purposes. This process can be quite expensive, so it is advised to adhere to the following pipeline:

from sklearn.datasets import fetch_20newsgroups
from sentence_transformers import SentenceTransformer
from bertopic import BERTopic
from umap import UMAP

# Prepare embeddings
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']
sentence_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = sentence_model.encode(docs, show_progress_bar=False)

# Train BERTopic
topic_model = BERTopic().fit(docs, embeddings)

# Run the visualization with the original embeddings
topic_model.visualize_documents(docs, embeddings=embeddings)

# Reduce dimensionality of embeddings, this step is optional but much faster to perform iteratively:
reduced_embeddings = UMAP(n_neighbors=10, n_components=2, min_dist=0.0, metric='cosine').fit_transform(embeddings)
topic_model.visualize_documents(docs, reduced_embeddings=reduced_embeddings)

**Comment by clarkegavin** [comment]
1. Visualise documents (visualize_documents) -> useful comparision between two datasets, but then only do this when we have selected the final model --> Done
2. Visualise Topic Hierarchy (visualize_hierarchy) -> can help to select an appropriate nr_topics -> not sure how relevant thsi will be... --> Done
3. Hierarchical Labels (hierarchical_topics -> visualize_hierarchy -> this is probably useful... -> Done
4. Text-based topic tree (get_topic_tree)
5. visualize hierarchical documents (hierarchical_topics -> visualize_hierarchical_documents) -> this could be useful as a comparison
6. Visualize terms (visualize_barchart) -> this is useful, again for comparison --> Done
7. Visualize topic similarity (visualize_heatmap()) - this is useful --> Done
8. Visualize term score decline (visualize_term_rank) - pass log_scale=True) -> Done
9. Visualize topics over time (topics_over_time) -> this is useful -> DONE
10. visualize topics per class (useful for TDs, Constituencies, Parties) (visualize_topics_per_class) - this is useful --> DONE
11. visualize probabilities or distribution (approximate_distribution
